pkgs/applications/virtualization/qemu/binfmt-p-wrapper.c

```diff
@@ -0,0 +1,91 @@
+// This is a tiny wrapper that converts the extra arv[0] argument
+// from binfmt-misc with the P flag enabled to QEMU parameters.
+// It also prevents LD_* environment variables from being applied
+// to QEMU itself.
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#ifndef TARGET_QEMU
+#error "Define TARGET_QEMU to be the path to the qemu-user binary (e.g., -DTARGET_QEMU=\"/full/path/to/qemu-riscv64\")"
+#endif
+
+extern char **environ;
+
+void realloc_ptr_array(char ***arr, size_t *alloc, size_t count) {
+    if (count > *alloc) {
+        *alloc = count;
+        *arr = (char**)realloc(*arr, (*alloc + 1) * sizeof(char**));
+
+        if (!*arr) {
+            fprintf(stderr, "FATAL: Failed to realloc\n");
+            abort();
+        }
+    }
+}
```

Comment:
The realloc things is quite complicated and messy here. The argument `alloc` is indeed "count" not "allocated", and also the alloc size should be multiples of `sizeof(char*)` not `sizeof(char**)` since `char**` is an array with elements `char*`.

Since we can know the length of new environ array is at most `len(environ)`, and the length of new argument array is at most:
```
[".../qemu-...", "-0", "argv0", "-E", "...=...", "--", "/path/to/elf", "args"]
3 + 2 * len(environ) + 2 + (argc - 3)
```
Simply `malloc` the upper bound once for both `new_environ` and `new_argv` would be enough.